core/chainio/avs_writer.go

```diff
@@ -91,3 +103,86 @@ func (w *AvsWriter) SendAggregatedResponse(batchMerkleRoot [32]byte, senderAddre
 
 	return &txHash, nil
 }
+
+func (w *AvsWriter) checkRespondToTaskFeeLimit(tx *types.Transaction, txOpts bind.TransactOpts, batchIdentifierHash [32]byte, senderAddress [20]byte) error {
+	aggregatorAddress := txOpts.From
+	simulatedCost := new(big.Int).Mul(new(big.Int).SetUint64(tx.Gas()), tx.GasPrice())
+	w.logger.Info("Simulated cost", "cost", simulatedCost)
+
+	// Get RespondToTaskFeeLimit
+	batchState, err := w.AvsContractBindings.ServiceManager.BatchesState(&bind.CallOpts{}, batchIdentifierHash)
+	if err != nil {
+		// Retry with fallback
+		batchState, err = w.AvsContractBindings.ServiceManagerFallback.BatchesState(&bind.CallOpts{}, batchIdentifierHash)
+		if err != nil {
+			// Fallback also failed, proceed with simulated cost checks
+			w.logger.Error("Failed to get batch state", "error", err)
+			w.logger.Info("Proceeding with simulated cost checks")
+			if err := w.compareAggregatorBalance(simulatedCost, aggregatorAddress); err != nil {
+				return err
+			}
+			if err := w.compareBatcherBalance(simulatedCost, senderAddress); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+
+	// At this point, batchState was successfully retrieved
+	// Checking values against RespondToTaskFeeLimit.
+
+	respondToTaskFeeLimit := batchState.RespondToTaskFeeLimit
+	w.logger.Info("Batch RespondToTaskFeeLimit", "RespondToTaskFeeLimit", respondToTaskFeeLimit)
+
+	if respondToTaskFeeLimit.Cmp(simulatedCost) < 0 {
+		return fmt.Errorf("cost of transaction is higher than Batch.RespondToTaskFeeLimit")
+	}
+	err = w.compareAggregatorBalance(respondToTaskFeeLimit, aggregatorAddress)
+	if err != nil {
+		return err
+	}
+	err = w.compareBatcherBalance(respondToTaskFeeLimit, senderAddress)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *AvsWriter) compareAggregatorBalance(amount *big.Int, aggregatorAddress common.Address) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// Get Agg wallet balance
+	aggregatorBalance, err := w.Client.BalanceAt(ctx, aggregatorAddress, nil)
+	if err != nil {
+		aggregatorBalance, err = w.ClientFallback.BalanceAt(ctx, aggregatorAddress, nil)
+		if err != nil {
+			// Ignore and continue.
+			w.logger.Error("failed to get aggregator balance: %v", err)
+			return nil
+		}
+	}
+	w.logger.Info("Aggregator balance", "balance", aggregatorBalance)
+	if aggregatorBalance.Cmp(amount) < 0 {
+		return fmt.Errorf("cost is higher than Aggregator balance")
+	}
+	return nil
+}
+
+func (w *AvsWriter) compareBatcherBalance(amount *big.Int, senderAddress [20]byte) error {
+	// Get batcher balance
```

Comment:
maybe this should get a context as well
```suggestion
func (w *AvsWriter) compareBatcherBalance(amount *big.Int, senderAddress [20]byte) error {
        ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	// Get batcher balance
```

Reply:
We are using the `ServiceManager` and `ServiceManagerFallback` for the `get BatchersBalances` call, which doesn't need a context to be passed to it (i'm guessing it is already configured at this point).
While in the other case we use `eth.Client`, which needs a context